### DISCUSSION TITLE
Session storage support in app.storage

### DISCUSSION THREAD

**Discussion by sklibanov312** [zauberzeug/nicegui#2378] [Q&A] [answered]
### Question

The document https://nicegui.io/documentation/storage#storage clearly explains that app.storage.browser values are shared among all the tabs in the browser. That's great for use cases that span different browser tabs, but that's not my use case.

I have a need to use session storage in my application. I need each browser tab to have its own context. This support will help me to minimize the user-facing impact of server restarts and page reloads by storing the user's inputs in the client. 

I looked for usage of sessionStorage in the nicegui libraries, both in the server and in the javascript, and haven't found anything. I've therefore written these functions to update and retrieve session storage values:

```py
async def _updateClientSessionStorage(client: nicegui.client.Client, updateDict: Dict[str, str], timeout=60):
    """
    :param client: The client to update. Use the .client property of any page element. Ensure the client is connected by awaiting client.connected() with a suitable timeout
    :param updateDict: Keys and values you want to store in session storage. These must be strings
    :param timeout: How long to wait for a response, in seconds. Passed directly into run_javascript timeout
    """
    encodedUpdateDict = json.dumps(updateDict)
    js = """
    for (const [key, value] of Object.entries(%s)) 
    {
        sessionStorage.setItem(key, value);
    }
    return "Done.";
    """ % (encodedUpdateDict,)
    # Without returning Done, there is an exception in the response handler...
    _result = await client.run_javascript(js, timeout=timeout)



async def _getClientSessionStorageValues(client: nicegui.client.Client, keys: List[str], timeout=60) -> Dict[str, str]:
    """
    :param client: The client to update. Use the .client property of any page element. Ensure the client is connected by awaiting client.connected() with a suitable timeout
    :param keys: The keys whose values you want to fetch from session storage
    :param timeout: How long to wait for a response, in seconds. Passed directly into run_javascript timeout
    :return: A dictionary containing the keys and values that were retrieved. Keys that are absent from client session storage are omitted.
    """
    encodedKeys = json.dumps(keys)
    js = """
    const result = {};
    for (const key of %s) {
        const value = sessionStorage.getItem(key);
        if (value !== null) {
            result[key] = value;
        }
    }
    return result;
    """ % (encodedKeys,)
    result = await client.run_javascript(js, timeout=timeout)
    return result

```

Are there any problems with this approach? 

Is it worth it to check that the result we're returning from the getter is indeed a dict from string to string? Or should we trust user code to make this verification?

Should I (or someone...) make a PR that adds this to app.storage.* or similar?

Thanks!

**Comment by rodja** [comment, marked as answer]
Oh that is interesting! This could be the solution for https://github.com/zauberzeug/nicegui/discussions/1308. Session Storage stays available for page reloads and navigation as long as the browser tab is not closed or domain is switched. Would you like to create a pull request?

**Reply by sklibanov312** [reply]
Let's move the conversation to #1308 .
